---
title: "Gene enrichment analysis"
output: html_document
---

```{r load-packages, include=FALSE}
#library(officedown)
library(officer)
library(flextable)
library(dplyr)
library(knitr)
library(magrittr)
library(knitr)
library(ggplot2)
library(data.table)
library(RColorBrewer)
library(jsonlite)
library(tidyr)
library(stringr)

knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
set_flextable_defaults(font.family = "Arial", font.size = 10, padding.top = 3, padding.bottom = 3)
```

```{r , echo=FALSE}
# Height for sample plots
# default is good for few samples and Read coverage plot
default_height <- 5
# custom is needed if several samples >= 50 ? -> 10
custom_height1 <- default_height
# another custom if the first one is not enough (rRNA plot can need higher value because of pair-end doubling the amount of samples) -> 15
custom_height2 <- default_height
```

```{r, results='asis', echo=FALSE}
#BASEDIR <- "/mnt/ssd/ssd_1/snakemake/"
#BASEDIR <- "/mnt/nfs/shared/CFBioinformatics/all_projects/"

DE_FC <- "DE_feature_count"
if(dir.exists(paste0("results","/DE_RSEM"))==T){
  var.rsem <- 1
  DE_RSEM <- "DE_RSEM"
}else{
   var.rsem <- 0
}

`%!in%` <- Negate(`%in%`)

```

# Acknowledgement
- We request co-authorship (and will assist with manuscript preparation) if we have developed novel tools, algorithms, or pipelines, participated in experiment design planning, or have contributed to a biological question addressed in a manuscript. We ask that, at a minimum, you acknowledge us in a publication to which we have contributed routine analysis without further support and discussions, data management or conversion, or data submission services.
- Core Facility Bioinformatics of CEITEC Masaryk University is gratefully acknowledged for the obtaining of the scientific data presented in this paper.

# Description
```{r, results='asis', echo=FALSE}
config_json = as.data.table(fromJSON(paste0(snakemake@params$config)))
config_json[,condition := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$condition)]
config_json[,tag := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$tag)]
config_json[,full_name := sapply(1:length(config_json$samples),function(x) config_json$samples[[x]]$sample_name)]
config_json[, species_name := gsub("\\)","",gsub(".*\\(","",species))]

```

- The main goal of the experiment is to perform Gene Enrichment analysis.
- We have used as input the results from Differential Expression analysis generated from DESeq2 (DESeq2.tsv).
In general, we have `r length(config_json[,full_name])` `r unique(config_json[,species_name]` samples. We have `r length(config_json[,unique(condition)])` conditions (`r config_json[,unique(condition)]`), those conditions are sequenced in `r round(length(config_json[,unique(condition)])/length(config_json[,full_name]))` replicates, which is sufficient enough to obtain statistical significance.
- Conditions to compare are:
`r border_remove(theme_booktabs(delete_part(qflextable(data = as.data.frame(comparison_dt$Conditions)),part = "header")))`

- For each condition we consider the Differential Expression analysis results obtained from
DESeq2 (DESeq2.tsv).

Reminder: **cond1_vs_cond2** compares cond1 to cond2. In the DE results then - positive
log2FC signalizes more expression in cond1 (or less in cond2) and negative log2FC signalizes
less expression in cond2 (or more in cond1).

## Agreed task(s)
Gene Enrichment and Gene Set Enrichment Analysis (GSEA):

`r if(config_json$onthology[1]==TRUE){paste0("- Gene Onthollogy (GO)")}`
`r if(config_json$kegg[1]==TRUE){paste0("- KEGG pathways")}`
`r if(config_json$reactome[1]==TRUE){paste0("- REACTOME pathways")}`
`r if(config_json$wikipathways[1]==TRUE){paste0("- WikiPathways (WP)")}`

## Analysis
- Gene Enrichment is based only on the genes for the DESeq2 result table, that are considered Differentially Expressed so that their **log2FoldChange is +/- `r config_json$cutoff_log2fc_enrich[1]`** and their **adjusted p-value is â‰¤ `r config_json$cutoff_padj_enrich[1]`**. Gene Enrichment will look if some feature (onthologies or pathways) are enriched among those DE genes.
- GSEA takes in consideration all the genes from the DESeq2 result table for which both pvalue and adjusted p-value exist (not NA). Then those genes are sorted according to their log2FoldChange. The enrichment is then performed and can detect features that are common for gene that were influenced similarly by the treatment condition.
- Both Enrichment and GSEA are performed using the R package **clusterProfiler**.
- Gene Onthology annotations are provided by the package **`r config_json$organism_go[1]`**.
- **KEGG**, **REACTOME** and **WikiPathways** annotations come from their respective databases (R package **ReactomePA** is needed to analyze REACTOME pathways)
- Complete settings, used commands, tool versions and methodology part for a publication can be provided upon request.

## Results
- For each analysis, a result table is generated containing the list of enriched features, an adjusted p-value and the list of the genes sharing those features. We used an adjusted p-value cutt-off of `r config_json$enrich_padj[1]` on the results.
- For Gene Enrichment we look for the `r config_json$n_up[1]` most enriched features.
- For GSEA we look for the `r config_json$n_up[1]` most enriched features among the up-regulated genes and the `r config_json$n_down[1]` most enriched features among the down-regulated genes.

### Gene Onthology
#### Biological Process


```{r enrich-BP, fig.cap = paste0("Enrichment GO Biological Process: onthologies are sorted by adjusted p-value with the onthology on the top of the graph having the smallest p-value."), echo=FALSE, out.width="80%", fig.align="center", results="asis", fig.show="hold", dev="svg"}

BP.files = snakemake@input$gobpE
BPg.files = snakemake@input$gobpG
MF.files = snakemake@input$gomfE
MFg.files = snakemake@input$gomfG
CC.files = snakemake@input$goccE
CCg.files = snakemake@input$goccG

#BP.files = list.files(path=paste0(comparison,"/",config_json$biotype[1],"/enrichment_GO"), pattern = "BP.pdf", full.names = TRUE)

if(length(BP.files)>1){
  for(i in 2:length(BP.files)){
    cat("![",paste0(gsub("/"," - ",gsub("results/","",comparison[i]))," - Enrichment GO Biological Process: onthologies are sorted by adjusted p-value with the onthology on the top of the graph having the smallest p-value."),"](",BP.files[i],")")
    cat('\n\n')
  }
}

cat("![",paste0(gsub("/"," - ",gsub("results/","",comparison[1]))," - Enrichment GO Biological Process: onthologies are sorted by adjusted p-value with the onthology on the top of the graph having the smallest p-value."),"](",BP.files[1],")")
#knitr::include_graphics(BP.files[1])

```

---